### ISSUE TITLE
[STORM-1737] storm-kafka-client has compilation errors with Apache Kafka 0.10

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5572]
<p>when compiled with Apache Kafka 0.10 branch getting following errors</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[ERROR] /Users/harsha/code/hwx/storm/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:[163,51] incompatible types: org.apache.kafka.common.TopicPartition cannot be converted to java.util.Collection<org.apache.kafka.common.TopicPartition>
[ERROR] /Users/harsha/code/hwx/storm/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:[166,45] incompatible types: org.apache.kafka.common.TopicPartition cannot be converted to java.util.Collection<org.apache.kafka.common.TopicPartition>
[ERROR] /Users/harsha/code/hwx/storm/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:[175,51] incompatible types: org.apache.kafka.common.TopicPartition cannot be converted to java.util.Collection<org.apache.kafka.common.TopicPartition>
[ERROR] /Users/harsha/code/hwx/storm/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:[177,45] incompatible types: org.apache.kafka.common.TopicPartition cannot be converted to java.util.Collection<org.apache.kafka.common.TopicPartition>
[ERROR] /Users/harsha/code/hwx/storm/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java:[252,41] incompatible types: org.apache.kafka.common.TopicPartition cannot be converted to java.util.Collection<org.apache.kafka.common.TopicPartition>
</pre>
</div></div>

---
<details><summary><i>Originally reported by <a title="Harsha" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1737" target="_blank">storm-kafka-client has compilation errors with Apache Kafka 0.10</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Hugo Da Cruz Louro" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Blocker
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-07-15T17:52:32+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p>The above issue due to the api change from new kafka consumer 0.9 to 0.10. Given that Kafka Consumer interface marked as unstable we should make this change on our side by updating the pom.xml kafka.version to 0.10 and make the required api changes in kafka spout.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>If this issue can be fixed by several days I'm OK to mark this as 'Blocker', but if not I don't think this issue should block releasing 1.0.2.<br/>
Yes Kafka is a kind of de-facto standard queue now so we should consider Kafka as first class, but Apache Kafka 0.10 was released just 1 month ago and no other streaming frameworks seem support Kafka 0.10.</p>

<p>Btw, can we make storm-kafka-client not using methods marked as 'Unstable'? Since it's marked as 'Unstable' Kafka community can change that at other minor updates, and I don't think we should follow their changes ASAP.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a> as of release 0.10.0 the @InterfaceStability.Unstable annotation was removed from the Consumer interface.</p>

<p>I am working on this now. I will a PR by EOD</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a> anyone using storm-kafka-client that built with 0.9 binaries will not  be able to use 0.10 release. Yes we don't need to follow kafka releases but in this case we should take the advantage of api stability that came in 0.10 release and fix the broken compatibility .</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user hmcl opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1737" title="storm-kafka-client has compilation errors with Apache Kafka 0.10" class="issue-link" data-issue-key="STORM-1737"><del>STORM-1737</del></a>: storm-kafka-client has compilation errors with Apache Kafka 0.10</p>



<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/hmcl/storm-apache" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/hmcl/storm-apache</a> <a href="https://issues.apache.org/jira/browse/STORM-1737" title="storm-kafka-client has compilation errors with Apache Kafka 0.10" class="issue-link" data-issue-key="STORM-1737"><del>STORM-1737</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1556</p>

<hr />
<p>commit 04f00e56beb9799dc8b8fc732271b786ced5c21e<br/>
Author: Hugo Louro <hmclouro@gmail.com><br/>
Date:   2016-04-29T16:18:05Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1737" title="storm-kafka-client has compilation errors with Apache Kafka 0.10" class="issue-link" data-issue-key="STORM-1737"><del>STORM-1737</del></a>: storm-kafka-client has compilation errors with Apache Kafka 0.10</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556#discussion_r70542939" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556#discussion_r70542939</a></p>

<p>    &#8212; Diff: external/storm-kafka/src/test/org/apache/storm/kafka/KafkaTestBroker.java &#8212;<br/>
    @@ -17,21 +17,21 @@<br/>
      */<br/>
     package org.apache.storm.kafka;</p>

<p>    +import org.apache.commons.io.FileUtils;<br/>
     import org.apache.curator.framework.CuratorFramework;<br/>
     import org.apache.curator.framework.CuratorFrameworkFactory;<br/>
     import org.apache.curator.framework.imps.CuratorFrameworkState;<br/>
     import org.apache.curator.retry.ExponentialBackoffRetry;<br/>
     import org.apache.curator.test.InstanceSpec;<br/>
     import org.apache.curator.test.TestingServer;</p>

<p>    -import kafka.server.KafkaConfig;<br/>
    -import kafka.server.KafkaServerStartable;<br/>
    -import org.apache.commons.io.FileUtils;<br/>
    -<br/>
     import java.io.File;<br/>
     import java.io.IOException;<br/>
     import java.util.Properties;</p>

<p>    +import kafka.server.KafkaConfig;<br/>
    +import kafka.server.KafkaServerStartable;<br/>
    +<br/>
     /**<br/>
    &#8212; End diff &#8211;</p>

<p>    Out of scope, but we need to get rid of this comment.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    What's our plan to support version lines of Kafka? I think we need to guide storm-kafka to users that for 0.8 you need to blabla, and for 0.9, and for 0.10.<br/>
    And storm-kafka-client is no longer compatible with 0.9, it would be better to go with <a href="https://issues.apache.org/jira/browse/STORM-1876" title="Provide option to build storm-kafka against different kafka clients" class="issue-link" data-issue-key="STORM-1876"><del>STORM-1876</del></a>(<a href="https://issues.apache.org/jira/browse/STORM-1876" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-1876</a>) (#1482). If this patch merged in without <a href="https://issues.apache.org/jira/browse/STORM-1876" title="Provide option to build storm-kafka against different kafka clients" class="issue-link" data-issue-key="STORM-1876"><del>STORM-1876</del></a>, 0.9 users should download the code, and *<b>change directory to external/storm-kafka</b>*, and build module with proper `kafka.version` and `kafka.artifact.id` parameters.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @HeartSaVioR 0.9 users won't be able to use the new API anyway. So why are we worried about that part. ). 0.9 users can still use storm-kafka but the new kafka consumer in storm-kafka-client</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @HeartSaVioR I don't think this patch need to wait on storm-kafka changes in <a href="https://issues.apache.org/jira/browse/STORM-1876" title="Provide option to build storm-kafka against different kafka clients" class="issue-link" data-issue-key="STORM-1876"><del>STORM-1876</del></a>. Irrespective both these patches storm-kafka will work against 0.8, 0.9 and 0.10</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @harshach <br/>
    Sorry I've confused with compatibility between client and broker version. Seems like kafka 0.9 broker is compatible with kafka 0.8 client. I guess you're based on this. Is it correct?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @HeartSaVioR Yes. Users can still use the storm-kafka connector (old api) to connect to any version of kafka. Given that this is new consumer api we better move to compatible API soon. Hence the reason to push this as part of 1.0.2 release.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @hmcl Please update README.md to describe this module is compatible with 0.10.0.</p>

<p>    @harshach We're having potential users which already uses storm-kafka-client with Kafka 0.9. It might be better for them to guide not updating their storm-kafka-client version as same as storm-core due to client version change, or dropping usage of storm-kafka-client and back to storm-kafka.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    And let's clarify: what I'm thinking about is not only API level compatibility but Kafka client - broker version compatibility. We modified storm-kafka version to 0.8 due to this issue.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @HeartSaVioR we released 1.0 like few weeks back and even they are using the new consumer api in 0.10 will be api compatible with 0.9 as well it just the parameter types that are changed. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @harshach just wanted to make sure that compatibility of kafka 0.10  with 0.9 broker has been tested. </p>

<p>    Also given that there are three version lines of kafka client and two different kafka modules within storm, we should clearly document which combinations of kafka-client are supported. </p>

<p>    e.g.<br/>
    storm-kafka supports kafka client 0.8.x and 0.9.x<br/>
    storm-kafka-client supports kafka client 0.10.x</p>

<p>    Apart from that kafka dependencies should be marked provided, so that users are forced to chose a kafka client version explicitly. Whether the kafka client version chosen by user works with broker, should be left for user to decide. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @abhishekagarwal87 We can put that in the README on which version we are going to support for storm-kafka-client.<br/>
    agree on making the dependencies provided.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    I will make the scope changes in #1482 </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @HeartSaVioR is this ok to be merged in.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    Where we would want to update support versions on storm-kafka-client? <br/>
    I'd be +1 when we're clear to where we handle that thing.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @HeartSaVioR As stated before we can add this to README of storm-kafka-client</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @hmcl can you update the README with a Note: supports from kaka-client 0.10 version on wards</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    Sorry to not clear on `where`: I mean which PR (or new issue) will handle updating the doc. Now I'm +1 once @hmcl updates the doc.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Removing fix version and adding to 1.0.2 epic.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user hmcl commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    @harshach @HeartSaVioR @abhishekagarwal87, thank you for your feedback. The updated ReadME, including the added Wildcard Support will be out in a few mins.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

<p>    +1.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz">ptgoetz</a>:</i>
<p>Merged to master and 1.x-branch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz">ptgoetz</a>:</i>
<p>Moving to 1.1.0 release epic.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1556</a></p>

**Comment by jira-importer** [comment]
<i>[Epic: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1856%20label%3Aepic">STORM-1856</a>]</i>
